### PR TITLE
Initial addition of riscv-trace-spec signals to ibex_core

### PR DESCRIPTION
**Pull request by timeprints**
The mandatory signals defined by the riscv-trace-spec are added to ibex_core outputs. 
They are also connected initially to some signals from the ibex core. 
These signals are to be checked and corrected as more understanding is gained.

### CODE REVIEW

File: rtl/ibex_core.sv [hunk 1]
```diff
@@ -95,6 +95,18 @@ module ibex_core #(
     output logic [ 3:0] rvfi_mem_wmask,
     output logic [31:0] rvfi_mem_rdata,
     output logic [31:0] rvfi_mem_wdata,
+
+`endif
+
+`ifdef RV_TRACE
+    // riscv_trace_spec signals
```
Comment (imphil) [inline rtl/ibex_core.sv:102]:
```suggestion
    // Trace port according to the RISC-V Trace Specification v1.0
```

File: rtl/ibex_core.sv [hunk 2]
```diff
@@ -95,6 +95,18 @@ module ibex_core #(
...
+
+`ifdef RV_TRACE
+    // riscv_trace_spec signals 
+    output logic [1:0]  rv_trace_ctype,
+    output logic [3:0]  rv_trace_itype,
+    output logic [31:0] rv_trace_iaddr,
+    output logic [11:0] rv_trace_cause,
+    output logic [31:0] rv_trace_tval,
+    output logic [1:0]  rv_trace_priv, 
+    output logic [0:0]  rv_trace_iretire,
```
Comment (imphil) [inline rtl/ibex_core.sv:109]:
```suggestion
    output logic        rv_trace_iretire,
```

File: rtl/ibex_core_tracing.sv [hunk 3]
```diff
@@ -67,6 +67,17 @@ module ibex_core_tracing #(
...
 
+`ifdef RV_TRACE
+    // riscv_trace_spec signals 
+    output logic [1:0]  rv_trace_ctype,
+    output logic [3:0]  rv_trace_itype,
+    output logic [31:0] rv_trace_iaddr,
+    output logic [11:0] rv_trace_cause,
+    output logic [31:0] rv_trace_tval,
+    output logic [1:0]  rv_trace_priv, 
+    output logic [0:0]  rv_trace_iretire,
```
Comment (imphil) [inline rtl/ibex_core_tracing.sv:78]:
```suggestion
    output logic        rv_trace_iretire,
```

File: doc/riscv_trace.rst [hunk 4]
```diff
@@ -0,0 +1,69 @@
+.. _riscv_trace:                                                                                         
+
+   :name: riscv-trace-spec
+   :align: center
```
Comment (imphil) [inline doc/riscv_trace.rst:4]: Remove these two lines.

File: doc/riscv_trace.rst [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+RISC-V Trace Specs Interface
```
Comment (imphil) [inline doc/riscv_trace.rst:6]:
```suggestion
RISC-V Trace Interface
```

File: doc/riscv_trace.rst [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+============================
+
+Ibex implements a trace interface compatible with the riscv-trace-spec Version 1.0 issued in March 20, 2020.
```
Comment (imphil) [inline doc/riscv_trace.rst:9]:
```suggestion
Ibex implements an instruction trace interface compatible with the `RISC-V Trace Specification, version 1.0, as released on March 20, 2020 <https://github.com/riscv/riscv-trace-spec/>`_. 
```

File: doc/riscv_trace.rst [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+Ibex implements a trace interface compatible with the riscv-trace-spec Version 1.0 issued in March 20, 2020. 
+This interface currently exports needed internal signal as ibex_core outputs to be used to encode the trace packets to be logged.
+Signals implemented until now are shown in the table below:
+
+
+Trace Signals
+-------------
+
+For now, only single retirement per block is implemented.
```
Comment (imphil) [inline doc/riscv_trace.rst:17]: Ibex doesn't retire more than one instruction per cycle, and it's unlikely that this will ever change (famous last words). So you can remove this sentence.

File: doc/riscv_trace.rst [hunk 8]
```diff
@@ -0,0 +1,69 @@
+.. _riscv_trace:                                                                                         
+
+   :name: riscv-trace-spec
+   :align: center
+
+RISC-V Trace Specs Interface
+============================
+
+Ibex implements a trace interface compatible with the riscv-trace-spec Version 1.0 issued in March 20, 2020. 
+This interface currently exports needed internal signal as ibex_core outputs to be used to encode the trace packets to be logged.
```
Comment (imphil) [inline doc/riscv_trace.rst:10]: Most of the time, you can remove words such as "currently" from the documentation. Everything can change, it is implied that the documentation refers to the current state of the implementation.

```suggestion
This interface exports needed internal signals as ibex_core outputs to be used to encode the trace packets to be logged.
```

File: doc/riscv_trace.rst [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+Signals implemented until now are shown in the table below:
```
Comment (imphil) [inline doc/riscv_trace.rst:11]: You can remove this sentence.

File: doc/riscv_trace.rst [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+
+Trace Signals
+-------------
+
+For now, only single retirement per block is implemented. 
+                                                                       
+
++-----------------------+--------------------------------------+-------------------------------------------------------------+
+|   Signal Name         |            # of bits                 |                         Description                         |
```
Comment (imphil) [inline doc/riscv_trace.rst:21]: To get the table to render properly and be more readable, you can make the columns more narrow. 

```suggestion
| Signal Name | # of bits | Description |
```

File: rtl/ibex_core.sv [hunk 11]
```diff
@@ -1291,4 +1304,24 @@ module ibex_core #(
 
 `endif
 
+`ifdef RV_TRACE
+  // rv trace specs signals dummy assingments, to be checked
+    assign rv_trace_ctype   = 2'b00 ; // 0: means don't report
+    case ( {rvfi_intr,rvfi_trap,branch_decision,perf_branch} )
+      4'b100x:    assign rv_trace_itype = 4'b0001;
```
Comment (imphil) [inline rtl/ibex_core.sv:1311]: Let's go for `itype_width_p == 3` as a starting point.

File: doc/riscv_trace.rst [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+Trace Signals
+-------------
+
++--------------------+---------------------+-------------------------------------------------------------+
+|   Signal Name      |   Number of bits    |                         Description                         |
++====================+=====================+=============================================================+
+| rv_trace_ctype     |           2         | 0: Don’t report;                                            |
+|                    |                     | 1: Report imprecisely;                                      |
+|                    |                     | 2: Report precisely;                                        |
+|                    |                     | 3: Report as asynchronous discontinuity.                    |
```
Comment (imphil) [inline doc/riscv_trace.rst:22]: We removed this signal.

File: doc/riscv_trace.rst [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_itype     |           3         | 0: Final instruction in block is none of other itype codes; |
+|                    |                     | 1: Exception. An exception that traps occurred              |
+|                    |                     | following the final retired instruction in the block;       |
+|                    |                     | 2: Interrupt. An interrupt that traps occurred              |
+|                    |                     | following the final retired instruction in the block;       |
+|                    |                     | 3: Exception or interrupt return;                           |
+|                    |                     | 4: Nontaken branch;                                         |
+|                    |                     | 5: Taken branch;                                            |
+|                    |                     | 6: Uninferable jump if itype_width_p is 3                   |
```
Comment (imphil) [inline doc/riscv_trace.rst:32]: Only document what we actually iplement:

```suggestion
|                    |                     | 6: Uninferable jump                 |
```

File: doc/riscv_trace.rst [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+This interface exports needed internal signals as ibex_core outputs to be used to encode the trace packets to be logged.
+
+Signals implemented until now are shown in the table below:
+
+
+Trace Signals
+-------------
+
++--------------------+---------------------+-------------------------------------------------------------+
+|   Signal Name      |   Number of bits    |                         Description                         |
```
Comment (imphil) [inline doc/riscv_trace.rst:17]: Remove all unnecessary whitespace around these fields to make them less wide and the source more readable.

"Number of bits" can be shortened to "Bits"

File: doc/riscv_trace.rst [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+|                    |                     | -The instruction retired immediately after a jump with a    |
+|                    |                     | target that cannot be inferredfrom the source code (also    |
+|                    |                     | referred to as the target or destination of the jump);      |
+|                    |                     | -Taken and nontaken branches;                               |
+|                    |                     | -Last instruction retired before an exception or interrupt; |
+|                    |                     | -First instruction retired following an exception or intert;|
+|                    |                     | -The last instruction retired before a privilege change;    |
+|                    |                     | -The first instruction retired following a privilege change.|
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_cause     |          12         | ucause/scause/mcause CSR                                    |
```
Comment (imphil) [inline doc/riscv_trace.rst:46]: There's no scause in Ibex.

File: doc/riscv_trace.rst [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+|                    |                     | 1: Exception. An exception that traps occurred              |
+|                    |                     | following the final retired instruction in the block;       |
+|                    |                     | 2: Interrupt. An interrupt that traps occurred              |
+|                    |                     | following the final retired instruction in the block;       |
+|                    |                     | 3: Exception or interrupt return;                           |
+|                    |                     | 4: Nontaken branch;                                         |
+|                    |                     | 5: Taken branch;                                            |
+|                    |                     | 6: Uninferable jump if itype_width_p is 3                   |
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_iaddr     |          32         | Contains the value of the address in the cases it needs     |
```
Comment (imphil) [inline doc/riscv_trace.rst:34]: iaddr is *always* reported, there's no trace filtering happened yet.

File: doc/riscv_trace.rst [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+|                    |                     | referred to as the target or destination of the jump);      |
+|                    |                     | -Taken and nontaken branches;                               |
+|                    |                     | -Last instruction retired before an exception or interrupt; |
+|                    |                     | -First instruction retired following an exception or intert;|
+|                    |                     | -The last instruction retired before a privilege change;    |
+|                    |                     | -The first instruction retired following a privilege change.|
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_cause     |          12         | ucause/scause/mcause CSR                                    |
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_tval      |          32         | Associated trap value, e.g. the faulting virtual address for|
```
Comment (imphil) [inline doc/riscv_trace.rst:48]: - Ibex doesn't implement S mode, so there's no stval.
- Ibex doesn't support virtual memory, so instead of virtual address, it's just address.

File: doc/riscv_trace.rst [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+|                    |                     | address exceptions, as written to the utval/stval/mtval CSR |
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_priv      |          2          | Privilege level for all instructions retired on this cycle. |
```
Comment (imphil) [inline doc/riscv_trace.rst:51]: You can simplify this; Ibex never retires more than one instruction per cycle.

File: doc/riscv_trace.rst [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_iretire   |          1          | Number of instructions retired in this block (0 or 1).      |                         
++--------------------+---------------------+-------------------------------------------------------------+
+
+Parameters
+----------
+
+Referring to table 7.1, here are the parameters specified in the the implemented interface:
```
Comment (imphil) [inline doc/riscv_trace.rst:59]:
```suggestion
Referring to Table 7.1 in the RISC-V Trace Specification v1.0, here are the parameters specified in the the implemented interface:
```

File: doc/riscv_trace.rst [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+|                    |                     | -The first instruction retired following a privilege change.|
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_cause     |          12         | ucause/scause/mcause CSR                                    |
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_tval      |          32         | Associated trap value, e.g. the faulting virtual address for|
+|                    |                     | address exceptions, as written to the utval/stval/mtval CSR |
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_priv      |          2          | Privilege level for all instructions retired on this cycle. |
++--------------------+---------------------+-------------------------------------------------------------+
+| rv_trace_iretire   |          1          | Number of instructions retired in this block (0 or 1).      |
```
Comment (imphil) [inline doc/riscv_trace.rst:53]: You can simplify this to say "The trace signals are valid (1), or not (0)."

File: rtl/ibex_core.sv [hunk 21]
```diff
@@ -1291,4 +1310,85 @@ module ibex_core #(
...
+      4'b010x:    rv_trace_itype = 3'd2;
+      4'b001x:    rv_trace_itype = 3'd3;
+      4'b001x:    rv_trace_itype = 3'd4;
+      4'b0010:    rv_trace_itype = 3'd5;
+      4'b0011:    rv_trace_itype = 3'd6;
+      default:    rv_trace_itype = 3'd0;
+    endcase
+  end
+
+  // Direct assignment which don't need staging
```
Comment (imphil) [inline rtl/ibex_core.sv:1340]: The only signals which don't need to be delayed are ones which are taken from the writeback stage/execute stage (3/2 pipeline stage config, respectively).

File: rtl/ibex_core.sv [hunk 22]
```diff
@@ -1291,4 +1310,85 @@ module ibex_core #(
...
+  assign rv_trace_cause   = { exc_cause , 5'b00000 , csr_save_cause }; // CHECK: do we need 12 bit? 6,7?
+  assign rv_trace_tval    = csr_mtval;
```
Comment (imphil) [inline rtl/ibex_core.sv:1342]: csr_mtval comes form the ID/EX stage, so it needs to go into the `if (instr_id_done) begin` block below.

File: rtl/ibex_core.sv [hunk 23]
```diff
@@ -1291,4 +1310,85 @@ module ibex_core #(
...
+      4'b001x:    rv_trace_itype = 3'd3;
+      4'b001x:    rv_trace_itype = 3'd4;
+      4'b0010:    rv_trace_itype = 3'd5;
+      4'b0011:    rv_trace_itype = 3'd6;
+      default:    rv_trace_itype = 3'd0;
+    endcase
+  end
+
+  // Direct assignment which don't need staging
+  assign rv_trace_cause   = { exc_cause , 5'b00000 , csr_save_cause }; // CHECK: do we need 12 bit? 6,7?
```
Comment (imphil) [inline rtl/ibex_core.sv:1341]: - exc_cause comes from the execute stage, so it needs to go into the `if (instr_id_done) begin` block below.
- Please specify this signal better in the signal description of the documentation, and then double-check that your implementation matches that description.

File: rtl/ibex_core.sv [hunk 24]
```diff
@@ -1291,4 +1310,85 @@ module ibex_core #(
 
 `endif
 
+`ifdef RV_TRACE
+  // rv trace specs signals dummy assingments, to be checked
+  always_comb begin
```
Comment (imphil) [inline rtl/ibex_core.sv:1315]: All these signals also need to be delayed as you've done for the other signals. Please figure out for each signal from which pipeline stage you got them from (all from the ID/EX stage?), and then delay them appropriately.

File: doc/riscv_trace.rst (hunk 16 above)
Comment (timeprints) [reply doc/riscv_trace.rst:34]: So you mean we will be selecting when to actually to report it later by the encoder, and here we just describe the interface in ibex-core?

Comment (imphil) [reply doc/riscv_trace.rst:34]: Yes, the interface you're describing here is the core interface; it's effectively the text form of the signals you're creating in `ibex_core.sv`.

File: rtl/ibex_core.sv [hunk 25]
```diff
@@ -1291,4 +1310,85 @@ module ibex_core #(
...
+      4'b0010:    rv_trace_itype = 3'd5;
+      4'b0011:    rv_trace_itype = 3'd6;
+      default:    rv_trace_itype = 3'd0;
+    endcase
+  end
+
+  // Direct assignment which don't need staging
+  assign rv_trace_cause   = { exc_cause , 5'b00000 , csr_save_cause }; // CHECK: do we need 12 bit? 6,7?
+  assign rv_trace_tval    = csr_mtval; 
+  assign rv_trace_priv    = priv_mode_id;
```
[the remark this answers is not in the document]
Comment (timeprints) [reply rtl/ibex_core.sv]: rvfi_mode is now timely aligned with rv_trace_priv, should it be delayed?

File: examples/simple_system/ibex_simple_system.core [hunk 26]
```diff
@@ -23,6 +23,10 @@ filesets:
 
 
 parameters:
+  RV_TRACE:
+    datatype: bool
+    paramtype: vlogdefine
```
Comment (imphil) [inline examples/simple_system/ibex_simple_system.core:28]: Please add a `description`.

File: ibex_core_tracing.core [hunk 27]
```diff
@@ -28,6 +28,10 @@ parameters:
     paramtype: vlogdefine
     default: true
 
+  RV_TRACE:
+    datatype: bool
+    paramtype: vlogdefine
```
Comment (imphil) [inline ibex_core_tracing.core:33]: Please add a `description` here as well.

File: rtl/ibex_core.sv [hunk 28]
```diff
@@ -1291,4 +1309,91 @@ module ibex_core #(
...
+  assign rv_trace_itype   = rv_trace_itype_stage   [RV_TRACE_STAGES-1];
+  assign rv_trace_cause   = { rv_exc_cause_stage   [RV_TRACE_STAGES-1] , csr_save_cause }; 
+  assign rv_trace_tval    = rv_trace_tval_stage    [RV_TRACE_STAGES-1];
+  assign rv_trace_priv    = priv_mode_id; 
+
+  // staging signals which need it
+  for (genvar i = 0;i < RV_TRACE_STAGES; i = i + 1) begin : g_rv_trace_stages
+    always_ff @(posedge clk or negedge rst_ni) begin
+      if (!rst_ni) begin
+        rv_trace_iaddr_stage[i]   <= 0;
```
Comment (imphil) [inline rtl/ibex_core.sv:1362]: You need to use either the exact bit width (e.g. `32'b0`), or use `'0` to fill the whole vector with 0s.

```suggestion
        rv_trace_iaddr_stage[i]   <= '0; 
```

File: rtl/ibex_core.sv [hunk 29]
```diff
@@ -1291,4 +1309,91 @@ module ibex_core #(
...
+    always_ff @(posedge clk or negedge rst_ni) begin
+      if (~rst_ni) begin
+        rv_trace_iretire <= 0;
+      end else begin
+        rv_trace_iretire_stage [RV_TRACE_STAGES-1] <= instr_done_wb;
+      end
+    end
+  end
+
+  assign rv_trace_iretire = rv_trace_iretire_stage [RV_TRACE_STAGES-1];
```
Comment (imphil) [inline rtl/ibex_core.sv:1351]: Here is the one and only place where you assign `rv_trace_iretire`. Remove the code you have above.

That's also where the CI failure is coming from, which reads:


```
%Error-BLKANDNBLK: ../src/lowrisc_ibex_ibex_core_tracing_0.1/rtl/ibex_core_tracing.sv:74:25: Unsupported: Blocked and non-blocking assignments to same variable: 'rv_trace_iretire'
   74 |     output logic        rv_trace_iretire,
      |                         ^~~~~~~~~~~~~~~~
```

File: rtl/ibex_core.sv [hunk 30]
```diff
@@ -1291,4 +1310,85 @@ module ibex_core #(
...
+  for (genvar i = 0;i < RV_TRACE_STAGES; i = i + 1) begin : g_rv_trace_stages
+    always_ff @(posedge clk or negedge rst_ni) begin
+      if (!rst_ni) begin
+        rv_trace_iaddr_stage[i]  <= 0; 
+        rv_trace_iretire_stage[i]<= 0;
+      end else begin
+        if (i == 0) begin
+          if(instr_id_done) begin
+            rv_trace_iaddr_stage[i]      <= {pc_id};
+            rv_trace_iretire_stage[i]    <= ex_valid;
```
[the remark this answers is not in the document]
Comment (imphil) [reply rtl/ibex_core.sv:1369]: @timeprints: Please have a look at this comment, it's important. Do *exactly* the same thing as the current code does to assign the various `rvfi_stage_valid`  and `rvfi_valid` signals.

File: rtl/ibex_core.sv [hunk 31]
```diff
@@ -1291,4 +1309,91 @@ module ibex_core #(
...
+      end
+    end
+  end
+
+  assign rv_trace_iretire = rv_trace_iretire_stage [RV_TRACE_STAGES-1];
+  assign rv_trace_iaddr   = rv_trace_iaddr_stage   [RV_TRACE_STAGES-1];
+  assign rv_trace_itype   = rv_trace_itype_stage   [RV_TRACE_STAGES-1];
+  assign rv_trace_cause   = { rv_exc_cause_stage   [RV_TRACE_STAGES-1] , csr_save_cause }; 
+  assign rv_trace_tval    = rv_trace_tval_stage    [RV_TRACE_STAGES-1];
+  assign rv_trace_priv    = priv_mode_id;
```
Comment (imphil) [inline rtl/ibex_core.sv:1356]: Please follow what is being done for `rvfi_stage_mode` and delay this signal as well.

**Comment by timeprints** [comment]
Should I also update ibex_simple_system.sv pins? They also result in the warning below for all rv_trace_* signals:
--
/home/massoud/tools/Tools/verilator/bin/verilator -f lowrisc_ibex_ibex_simple_system_0.vc --trace --trace-fst --trace-structs --trace-params --trace-max-array 1024 -CFLAGS "-std=c++11 -Wall -DVM_TRACE_FMT_FST -DTOPLEVEL_NAME=ibex_simple_system -g -O0" -LDFLAGS "-pthread -lutil -lelf" -Wall -Wno-PINCONNECTEMPTY -Wno-fatal --unroll-count 72
%Warning-PINMISSING: ../src/lowrisc_ibex_ibex_simple_system_0/rtl/ibex_simple_system.sv:160:7: Cell has missing pin: 'rv_trace_itype'
  160 |     ) u_core (
      |       ^~~~~~
                     ... Use "/* verilator lint_off PINMISSING */" and lint_on around source to disable this message.

**Comment by imphil** [comment]
> Should I also update ibex_simple_system.sv pins?

Yes, please add the pins to `ibex_simple_system.sv` in the instantiation, but don't connect them; https://github.com/lowRISC/ibex/blob/fb2a17a5a4fa917fda40619b9bdd25826dda1329/examples/simple_system/rtl/ibex_simple_system.sv#L147-L

something like that should do the trick
```systemverilog
      .rv_trace_iretire           (),
```

Make sure to enclose these port with the RV_TRACE define, as you do in the port list of the module itself. You can double-check if it worked by running fusesoc without the `--RV_TRACE` flag.

**Comment by timeprints** [comment]
A question: should this PR include the rest of implementation of itype (3 and 6)? or we limit it to the currently implemented itypes?
The current itype values of 1 and 2 are also still not at all tested. 

Here a two suggestions to implement the missing 2 itype (3 and 6):
First - To have another simple_sytem_with_inter_and_excep to be able to test the feature. 
Second - To implement the assignment of rv_trace_itype to value 3 (exception or interruption return), a suggestion is to do so by inspecting the current and previous rv_trace_iaddr and rv_trace_itype values, and when the previous itype is 1 or 2 (inter or excep), and the difference between the iaddr and the prev iaddr indicates a jump, which is not a taken branch, then itype would be 3. And itype=6 would be similarly assigned when the iaddr differences cannot be explained by any of the previous types. 

So, should we include these in the current PR?

**Comment by vogelpi** [comment]
Hey @imphil ,
do you know about the status of this PR?

**Comment by imphil** [comment]
This PR still needs a bit of work and there's currently no user for it. So let's close it and dig it out if we look at tracing again.
